### DISCUSSION TITLE
Can this be used to build expressions as well as queries?

### DISCUSSION THREAD

**Discussion by syonip** [react-querybuilder/react-querybuilder#405] [General]
I'm looking for a visual tool for building expressions that return a value that is not necessarily true or false.
So it's basically this library, plus unary operators.
Any way to add dynamic unary operators e.g. Math.ABS, String.toLower, or arbitrary function?
Also is there a way to make the top level return one value instead of AND/OR?

Any suggestions would be appreciated.
Thanks.

**Comment by jakeboone02** [comment]
Hi @syonip, you can replace any of the default components with your own, so theoretically yes, RQB could be used for that. Do you have a visual representation of what you'd like to see? A sketch or mockup or something? If I saw exactly what you were going for I could give you some tips on how to accomplish it.

**Comment by syonip** [comment]
Thanks @jakeboone02 for the quick response.
I want to build something like Telerik's Expression Editor: https://docs.telerik.com/devtools/winforms/controls/gridview/expression-editor/expression-editor

They have a windows component but no web version.
Thanks!

**Reply by jakeboone02** [reply]
I've been thinking about this a little bit and I can't think of a way to _properly_ support expression building within the current RQB structure. You can get part of the way, but not all the way. RQB is really designed to form questions, not expressions.

Populating the field list with functions, unary operators, etc., is fairly straightforward, as is creating a custom, dynamic `ValueEditor` that supports whatever "field" is selected. The problem is the nature of the hierarchy. RQB supports nested rules and groups, of course. You could populate the combinator selector with functions like `ROUND`, `MIN`, `MAX`, etc., to represent a function taking a complex expression as its argument, but then there's no way to place an operator like `+` or `-` between the "rules" (expressions) within the group. You could do that with `independentCombinators`, but then you'd lose the combinator selector at the group level.

Theoretically you could turn on `independentCombinators`, add a custom property to the group objects, read that property from a custom component in the group header (maybe augmenting the "add rule" button) that displays its own drop-down list for functions.

Could you provide an example of the kind of expression you want to be able to build?

**Reply by syonip** [reply]
So something like 
"IIF(TableFieldX == 10, TableFieldX,Round(TableFieldY))

**Reply by jakeboone02** [reply]
Right, so this is a single function `IIF`, with a specific signature. I can imagine mapping its three parameters to RQB "rules" so they can be complex themselves (like `Round(TableFieldY)`), but I'm curious what _other_ functions might be used, and how they'd be combined (addition, subtraction, etc.) or nested.

Will it always be a numeric result? If so, that limits the combinations to math operators like `+` and `-`, but even mathematical expressions are way too flexible to be fully represented in RQB, even with a lot of customization.

At this point I'm thinking the RQB structure is a little too restrictive for your use case.

**Comment by syonip** [comment]
Yeah it might require forking and changing the code a little bit. I'm also not sure about the complete requirements yet.
But thanks a lot for the advice!
